### DISCUSSION TITLE
Support for alternative serializers is great!

### DISCUSSION THREAD

**Discussion by nkosi23** [dotnet/orleans#9172] [General]
One thing that I hadn't realize was that Orleans is able to use System.Text.Json (and Newtonsoft) side-to-side with the native high-performance serializer. 

This is a game changer and removes a lot of friction especially for folks like me using Orleans from F# where we do not have the roslyn powered source-generation tools. 

While using System.Text.Json makes us lose some powerful features, such as object reference preservation, it is a great tradeoff to get development velocity. Especially since JsonConverters & Cie are typically already configured for the web api layer, so this makes passing rich messages to Orleans frictionless. And in F#, since objects are immutable, functions are built on the assumptions that we are working with copies of the data, not with a mutable reference.

Just wanted to thank you and highlight how massively useful this feature is. I'd be curious to know how much of a performance hit using System.Text.Json is compared to the native serializer, but I suspect that this will not make much difference for us anyway until we reach quite a large scale.

**Comment by nkosi23** [comment]
In case other people are curious regarding the performance implications of using System.Text.Json as the serializer, @ReubenBond discusses it in this video:  https://www.youtube.com/live/kgRag4E6b4c&t=1617 timestamp 26:57.

The built-in serializer is about one one-order of magnitude faster most of the time, which is quite a meaningful performance difference (even though we are talking about nanoseconds here). 

As far as we are concerned, we'll only reach out to the native serializer only on performance critical paths where we expect serialization to be quite intensive, or when we reach such a large scale that these nanoseconds start to move the needle as they add up. We're far from being at that place though :smile: 

Just putting this information here in case it is useful, in particular for folks using Orleans from F#, since the convenience of using System.Text.Json is important since we do not have code generators & Cie.

**Reply by ReubenBond** [reply]
Thank you for the kind words, @nkosi23, and apologies for not responding sooner. As for delegating to other serializers, there are likely ways to reduce the perf hit, and improvements there would be welcome contributions.

**Comment by gfix** [comment]
The F# support has become a bit better since the release of these PRs: #8739, #9039 and #9095 (all included since Orleans 9.0.0). You still need a C# project to be the target for the Orleans code generator. Also, since the F# compiler emits some internal members, you might need to make the projects you want the Orleans code generator to generate code for visible to the project being target for code generation via
```
[<InternalsVisibleTo("<the name of your C# project>")>]
do ()
```
